File: SLR/code/slr-demo.R

# STAT 331 - Simple Linear Regression Demo

dat <- read.csv("dataset/florange.csv")

head(dat)
attach(dat)

# Scatterplot
plot(acres, boxes)

# Summary statistics calculation examples
r <- cor(acres, boxes)
xbar <- mean(acres)
ybar <- mean(boxes)
sd_x <- sd(acres) 
sd_y <- sd(boxes) 

# Manual Calculation Examples
Sxx <- sum( (acres - xbar)^2 )
Sxy <- sum( (acres - xbar) * (boxes - ybar))

# R's "lm" function fits linear models
lm.1 <- lm(boxes ~ acres)
summary(lm.1) # recall RSE = estimate of sigma-hat

# Fitted values
lm.1$fitted.values

# Residuals
lm.1$residuals

# Manual calculation of sigm^2 estimate
sum(lm.1$residuals^2) / 23
# or sigma estimate: rse
sqrt(sum(lm.1$residuals^2) / 23)

# t-distribution values
qt(0.975, 23)
(1 - pt(17.263, 23)) * 2


# Discussion:
# - is sigma the same for all values of y? It appears not to be --> this is testing the equal variance or homoestadastic assumption. This appears to be violated so we can consider taking the log transformation

# - Are the error terms plausibly independent? (e.g., does know one e_i help predict e_j for a different county?)

File: Linear-Regression/code/slr-demo.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats
from sklearn.linear_model import LinearRegression
import statsmodels.api as sm

# STAT 331 - Simple Linear Regression Demo

# Read data
dat = pd.read_csv("dataset/florange.csv")

print(dat.head())

# Scatterplot
plt.figure(figsize=(8, 6))
plt.scatter(dat['acres'], dat['boxes'])
plt.xlabel('acres')
plt.ylabel('boxes')
plt.title('Scatterplot of acres vs boxes')
plt.show()

# Summary statistics calculation examples
r = dat['acres'].corr(dat['boxes'])
xbar = dat['acres'].mean()
ybar = dat['boxes'].mean()
sd_x = dat['acres'].std()
sd_y = dat['boxes'].std()

print(f"Correlation (r): {r:.4f}")
print(f"Mean of acres (xbar): {xbar:.4f}")
print(f"Mean of boxes (ybar): {ybar:.4f}")
print(f"Standard deviation of acres: {sd_x:.4f}")
print(f"Standard deviation of boxes: {sd_y:.4f}")

# Manual Calculation Examples
Sxx = ((dat['acres'] - xbar) ** 2).sum()
Sxy = ((dat['acres'] - xbar) * (dat['boxes'] - ybar)).sum()

print(f"Sxx: {Sxx:.4f}")
print(f"Sxy: {Sxy:.4f}")

# Using scikit-learn for linear regression
X = dat[['acres']]  # Feature matrix
y = dat['boxes']    # Target variable

lm_1 = LinearRegression()
lm_1.fit(X, y)

print(f"Intercept: {lm_1.intercept_:.4f}")
print(f"Coefficient: {lm_1.coef_[0]:.4f}")

# Using statsmodels for more detailed summary (similar to R's summary())
X_sm = sm.add_constant(dat['acres'])  # Adds intercept term
model = sm.OLS(dat['boxes'], X_sm)
results = model.fit()
print(results.summary())

# Fitted values
fitted_values = lm_1.predict(X)
print("Fitted values:", fitted_values[:5])  # Show first 5

# Residuals
residuals = dat['boxes'] - fitted_values
print("Residuals:", residuals[:5])  # Show first 5

# Manual calculation of sigma^2 estimate
n = len(dat)
p = 1  # number of predictors
df = n - p - 1  # degrees of freedom = 23 (as in R code)
sigma_sq_hat = (residuals ** 2).sum() / df
rse = np.sqrt(sigma_sq_hat)

print(f"Sigma^2 estimate: {sigma_sq_hat:.4f}")
print(f"RSE: {rse:.4f}")

# t-distribution values
t_value = stats.t.ppf(0.975, 23)
print(f"t-value for 95% confidence with 23 df: {t_value:.4f}")

# For the p-value calculation (equivalent to (1 - pt(17.263, 23)) * 2)
t_stat = 17.263
p_value = (1 - stats.t.cdf(t_stat, 23)) * 2
print(f"p-value for t-statistic 17.263: {p_value:.6f}")

# Additional useful calculations
# R-squared
r_squared = lm_1.score(X, y)
print(f"R-squared: {r_squared:.4f}")

# Correlation matrix
corr_matrix = dat[['acres', 'boxes']].corr()
print("Correlation matrix:")
print(corr_matrix)






